feat: add cakephp project type, fixes #5713

ajibarra opened the pull request:
<!-- 
  PR titles have very precise rules, please read 
  https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-pull-requests
-->

## The Issue

CakePHP projects must use  generic `php` type which does not allow automatic configuration of settings, access to cake shell through command, etc

## How This PR Solves The Issue

This PR adds cakephp project type to DDEV

## Manual Testing Instructions

```bash
    mkdir my-cakephp-app
    cd my-cakephp-app
    ddev config --project-type=cakephp --docroot=webroot --create-docroot --php-version=8.3
    ddev composer create--prefer-dist cakephp/app:~5.0
    ddev composer install
    ddev cake
    ddev launch
```

## Automated Testing Overview

* Added `cakephp` app type to `pkg/ddevapp/ddevapp_test.go`
* Added `cake` command to `cmd/ddev/cmd/commands_test.go`

## Related Issue Link(s)

#5713 

## Release/Deployment Notes

* CakePHP test app: https://github.com/ddev/test-cakephp

rfay commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -333,6 +333,22 @@ var (
+		// 18: CakePHP
+		{
+			Name:                          "TestPkgCakePHP",
+			SourceURL:                     "https://github.com/CakeDC/test-cakephp/archive/refs/tags/5.0.1.1.tar.gz",
```
We'll move this to[ a repo in ddev](https://github.com/ddev/test-cakephp) that can be maintained, updated, etc. I have given you maintainer privileges on that repo.

ajibarra replied:
Thank you, I will push the code to this repo and update devapp_test.go

ajibarra replied:
Moved!

rfay reviewed (commented):
So glad to have you working on this! 

Because you're a first-time contributor, each push has to have the github tests allowed by me.  The buildkite tests run automatically anyway, but you'll want the power to restart them. PM me (Discord or whatever) your email and I'll give you privs on buildkite.

ajibarra commented on docs/content/users/quickstart.md:
````diff
@@ -47,6 +47,36 @@ ddev launch
+=== "Composer"
+
+    ```bash
+    composer create-project --prefer-dist cakephp/app:~5.0 my-cakephp-app
````
In this case this is an step done before ddev project is created. However I understand your point, I will change readme to include New Project / Existing project structure as other types.

ajibarra replied:
Fixed!

ajibarra commented:
> So glad to have you working on this!
> 
> Because you're a first-time contributor, each push has to have the github tests allowed by me. The buildkite tests run automatically anyway, but you'll want the power to restart them. PM me (Discord or whatever) your email and I'll give you privs on buildkite.

I have sent you a PM on Discord. Thank you!

tyler36 commented on cmd/ddev/cmd/commands_test.go:
```diff
@@ -276,6 +276,17 @@ func TestCustomCommands(t *testing.T) {
 		assert.NoError(err)
 	}
 
+	// CakePHP commands should only be available for type laravel
```
```suggestion
	// CakePHP commands should only be available for type "cake"
```

ajibarra replied:
Fixed! Thank you!

rfay commented:
* https://github.com/ddev/test-cakephp/issues/1

ajibarra commented:
I would appreciate if you can take a look to issues in buildkite..Please let me know if you know what could be failing.

Additionally I have noticed if I do `ddev config` and go throught the steps, the post config action is called once (expected). But if I execute: `ddev config --project-type=cakephp --docroot=webroot --php-version=8.3`. The post config action is called twice. Checking the trace it looks it is called twice from the same place (see trace). It looks like a bug but I wanted to confirm with you.

```
runtime/debug.Stack()
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/runtime/debug/stack.go:24 +0x64
github.com/ddev/ddev/pkg/ddevapp.cakephpPostConfigAction(0x14000186708)
	/Users/ajibarra/work/ddev/pkg/ddevapp/cakephp.go:22 +0x28
github.com/ddev/ddev/pkg/ddevapp.(*DdevApp).PostConfigAction(0x14000186708)
	/Users/ajibarra/work/ddev/pkg/ddevapp/apptypes.go:420 +0xa4
github.com/ddev/ddev/pkg/ddevapp.(*DdevApp).WriteConfig(0x14000390708)
	/Users/ajibarra/work/ddev/pkg/ddevapp/config.go:248 +0x530
github.com/ddev/ddev/cmd/ddev/cmd.handleMainConfigArgs(0x10547da40, {0x104cbe93c?, 0xa?, 0x0?}, 0x14000390708)
	/Users/ajibarra/work/ddev/cmd/ddev/cmd/config.go:708 +0x21dc
github.com/ddev/ddev/cmd/ddev/cmd.handleConfigRun(0x10547da40, {0x140003e17a0, 0x0, 0x3})
	/Users/ajibarra/work/ddev/cmd/ddev/cmd/config.go:175 +0x1f8
github.com/spf13/cobra.(*Command).execute(0x10547da40, {0x140003e16b0, 0x3, 0x3})
	/Users/ajibarra/work/ddev/vendor/github.com/spf13/cobra/command.go:987 +0x828
github.com/spf13/cobra.(*Command).ExecuteC(0x105484a60)
	/Users/ajibarra/work/ddev/vendor/github.com/spf13/cobra/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/ajibarra/work/ddev/vendor/github.com/spf13/cobra/command.go:1039
github.com/ddev/ddev/cmd/ddev/cmd.Execute()
	/Users/ajibarra/work/ddev/cmd/ddev/cmd/root.go:162 +0x24
main.main()
	/Users/ajibarra/work/ddev/cmd/ddev/main.go:27 +0xb8

CakePHP: .env file exists already. Creating .env.ddev. You can rename it or copy settings to your .env file.
config/.env.ddev
runtime/debug.Stack()
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/runtime/debug/stack.go:24 +0x64
github.com/ddev/ddev/pkg/ddevapp.cakephpPostConfigAction(0x14000186e08)
	/Users/ajibarra/work/ddev/pkg/ddevapp/cakephp.go:22 +0x28
github.com/ddev/ddev/pkg/ddevapp.(*DdevApp).PostConfigAction(0x14000186e08)
	/Users/ajibarra/work/ddev/pkg/ddevapp/apptypes.go:420 +0xa4
github.com/ddev/ddev/pkg/ddevapp.(*DdevApp).WriteConfig(0x14000390708)
	/Users/ajibarra/work/ddev/pkg/ddevapp/config.go:248 +0x530
github.com/ddev/ddev/cmd/ddev/cmd.handleConfigRun(0x10547da40, {0x140003e17a0, 0x0, 0x3})
	/Users/ajibarra/work/ddev/cmd/ddev/cmd/config.go:187 +0x27c
github.com/spf13/cobra.(*Command).execute(0x10547da40, {0x140003e16b0, 0x3, 0x3})
	/Users/ajibarra/work/ddev/vendor/github.com/spf13/cobra/command.go:987 +0x828
github.com/spf13/cobra.(*Command).ExecuteC(0x105484a60)
	/Users/ajibarra/work/ddev/vendor/github.com/spf13/cobra/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/ajibarra/work/ddev/vendor/github.com/spf13/cobra/command.go:1039
github.com/ddev/ddev/cmd/ddev/cmd.Execute()
	/Users/ajibarra/work/ddev/cmd/ddev/cmd/root.go:162 +0x24
main.main()
	/Users/ajibarra/work/ddev/cmd/ddev/main.go:27 +0xb8
```

rfay commented on docs/content/users/quickstart.md:
```diff
@@ -47,6 +47,36 @@ ddev launch
+    mkdir my-cakephp-app
+    cd my-cakephp-app
+    ddev config --project-type=cakephp --docroot=webroot --create-docroot --php-version=8.3
+    ddev composer create-project --prefer-dist cakephp/app:~5.0 my-cakephp-app
```
I note that this isn't correct, so this wasn't walked through, please make sure to walk through and check it for validity, thanks!

```suggestion
    ddev composer create --prefer-dist cakephp/app:~5.0
```

ajibarra replied:
Thank you! it was left before changing to ddev composer. fixed

rfay commented on docs/content/users/quickstart.md:
````diff
@@ -47,6 +47,36 @@ ddev launch
+    ```bash
+    mkdir my-cakephp-app
+    cd my-cakephp-app
+    ddev config --project-type=cakephp --docroot=webroot --create-docroot --php-version=8.3
````
`--create-docroot` is no longer needed

```suggestion
    ddev config --project-type=cakephp --docroot=webroot --php-version=8.3
```

ajibarra replied:
Thank you! fixed!

rfay reviewed (changes requested):
Just a quick run-through to take a look as you requested at the post-config action. 

I note that it's not creating settings successfully yet I guess. 

```
CakePHP is NOT able to connect to the database.
Connection to Mysql could not be established: SQLSTATE[HY000] [2002] No such file or directory
``

rfay commented:
Thanks @ajibarra for the good eye in https://github.com/ddev/ddev/pull/5888#issuecomment-1972943332

I opened 
* https://github.com/ddev/ddev/pull/5923

and would love to have you take a look and manually test. 

I see that PostConfigAction has been very lightly used (only for Python projects) so doesn't have many miles on it, but sure glad to have you notice this.

I think you may want to consider looking at how Craft does this, which is in the post-start action I think. It may be more robust there?

rfay commented:
Oh and thank you so much for working on this! We sure welcome the CakePHP community!

ajibarra commented:
> Thanks @ajibarra for the good eye in [#5888 (comment)](https://github.com/ddev/ddev/pull/5888#issuecomment-1972943332)
> 
> I opened
> 
> * [fix: don't call app.WriteConfig twice #5923](https://github.com/ddev/ddev/pull/5923)
> 
> and would love to have you take a look and manually test.
> 
> I see that PostConfigAction has been very lightly used (only for Python projects) so doesn't have many miles on it, but sure glad to have you notice this.
> 
> I think you may want to consider looking at how Craft does this, which is in the post-start action I think. It may be more robust there?

Yeah I have switched to postStartAction. Initially I only want this to run once but then after testing the `config` BEFORE the project is actually created, it makes sense to always check in postStart.

ajibarra commented:
> Just a quick run-through to take a look as you requested at the post-config action.
> 
> I note that it's not creating settings successfully yet I guess.
> 
> ```
> CakePHP is NOT able to connect to the database.
> Connection to Mysql could not be established: SQLSTATE[HY000] [2002] No such file or directory
> ``
> ```

I have refactored the cakephp.go to ensure .env file is being loaded correctly so if you fetch the latest artifacts you should be able to smoothly go through the process (hopefully)

ajibarra commented:
I still have the issue with buildkite that I dont understand why it fails..

rfay commented:
> I still have the issue with buildkite that I dont understand why it fails..

Are you able to run the test locally? https://ddev.readthedocs.io/en/latest/developers/building-contributing/#testing

`GOTEST_SHORT=18 make test TESTARGS="-run TestDdevFullSiteSetup"`

Better yet, step through in a debugger and study the site that it builds before it starts testing it (or before it takes it down).

ajibarra commented:
> > I still have the issue with buildkite that I dont understand why it fails..
> 
> Are you able to run the test locally? https://ddev.readthedocs.io/en/latest/developers/building-contributing/#testing
> 
> `GOTEST_SHORT=18 make test TESTARGS="-run TestDdevFullSiteSetup"`
> 
> Better yet, step through in a debugger and study the site that it builds before it starts testing it (or before it takes it down).

yeah it runs successfully..I will try with the debugger. thanks for the suggestion.

rfay commented:
You may want to delete your cached assets in ~/.ddev/testcache, in case something there doesn't match what you actually have on test-cakephp.

Note that it's not just the buildkite tests that are failing, everything fails on TestAppTypeDetection, you'll be able to figure that one out super quickly.

ajibarra commented:
@rfay Finally! I was missing the cakephp folder in `pkg/ddevapp/testdata/TestApptypeDetection`

rfay commented on docs/content/users/quickstart.md:
```diff
@@ -47,6 +47,36 @@ ddev launch
+    cd my-cakephp-app
+    ddev config --project-type=cakephp --docroot=webroot --php-version=8.3
+    ddev composer create --prefer-dist cakephp/app:~5.0
+    ddev composer install
```
This isn't needed is it?  The `ddev composer create` does all the work?

```suggestion
```

rfay commented on docs/content/users/quickstart.md:
```diff
@@ -47,6 +47,36 @@ ddev launch
+    ddev config --project-type=cakephp --docroot=webroot --php-version=8.3
+    ddev start
+    ddev composer install
+    ddev cake
```
As a CakePHP n00b I found this confusing, because of course it just lists available commands. I am pretty sure you're just demonstrating that the command is there though.

ajibarra replied:
Yeah so people know it is available instead of doing ssh or ddev exec.

rfay reviewed (changes requested):
Thanks, this is looking great, congratulations on getting through all the test stuff. Some minor comments and suggestions.

ajibarra commented on pkg/ddevapp/cakephp.go:
```diff
@@ -0,0 +1,103 @@
+		"APP_DEFAULT_LOCALE":          "en_US",
+		"DATABASE_URL":                dbConnection + "://db:db@db:" + port + "/db",
+		"EMAIL_TRANSPORT_DEFAULT_URL": "smtp://localhost:1025",
+		"SECURITY_SALT":               generateRandomString(32),
```
Changes to util.RandString(32). Thank you!

ajibarra commented on pkg/ddevapp/cakephp.go:
```diff
@@ -0,0 +1,103 @@
+	return nil
+}
+
+func generateRandomString(length int) string {
+	b := make([]byte, length)
+	_, err := rand.Read(b)
+	if err != nil {
+		panic(err)
+	}
+	return base64.StdEncoding.EncodeToString(b)
+}
```
Changed to util.HashSalt(app.getName()). Thank you!

ajibarra commented:
@rfay  Hi, I am sorry for the delay but I was out of the office the last week. I have addressed all the comments and I have also improved a couple of things in .env file write process. Please let me know if something else is needed here. Thank you!

rfay commented:
SO GREEN!

stasadev commented on pkg/ddevapp/cakephp.go:
```diff
@@ -0,0 +1,93 @@
+		"export DATABASE_URL":                dbConnection + "://db:db@db:" + port + "/db",
+		"export EMAIL_TRANSPORT_DEFAULT_URL": "smtp://localhost:1025",
+		"export SECURITY_SALT":               util.HashSalt(app.GetName()),
+		"export DEBUGKIT_SAFETLD":            "site",
```
I wonder if this setting works?

When I open a project from the quickstart, I see:

Try adding your current top level domain to the [DebugKit.safeTld](https://book.cakephp.org/debugkit/5/en/index.html#configuration) config and reload.

![image](https://github.com/ddev/ddev/assets/24270994/d1b506e0-9ff7-4224-ba2e-b8b305a1a056)

ajibarra replied:
Yes! That's a great question. Currently DebugKit tld is only available to configure directly in the app. 

However discussing internally we thought it should definitely be configured using env variables as well so I'm preparing a pull request for DebugKit project to allow this. 

Thanks for your comments and feel free to raise any other concern!

stasadev replied:
Thank you for the explanation, I don't mind, even if it hasn't been added to the upstream yet.

tyler36 replied:
Confirming. I see all green (including database!) except for DebugKit which appears to be for local DEV setup.

![image](https://github.com/ddev/ddev/assets/7234392/2643ffab-58ec-49c2-9187-af8d18825c0a)

stasadev reviewed (approved):
Looks good to me, thank you!

rfay commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -334,6 +334,22 @@ var (
+			Name:                          "TestPkgCakePHP",
+			SourceURL:                     "https://github.com/ddev/test-cakephp/archive/refs/tags/5.0.1.1.tar.gz",
+			ArchiveInternalExtractionPath: "test-cakephp-5.0.1.1/",
+			FilesTarballURL:               "",
```
We don't need to test for existing static file?

ajibarra replied:
Updated test app to include a static page and updated test case

rfay commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -334,6 +334,22 @@ var (
+			FullSiteTarballURL:            "",
+			Type:                          nodeps.AppTypeCakePHP,
+			Docroot:                       "webroot",
+			Safe200URIWithExpectation:     testcommon.URIWithExpect{URI: "/", Expect: "Welcome to CakePHP"},
```
This is intended to be a static URI, file, html, whatever.

ajibarra replied:
Changed to `testcommon.URIWithExpect{URI: "/static/page.html", Expect: "This is a static page"},`

rfay commented on pkg/ddevapp/ddevapp_test.go:
```diff
@@ -334,6 +334,22 @@ var (
+		// 18: CakePHP
+		{
+			Name:                          "TestPkgCakePHP",
+			SourceURL:                     "https://github.com/ddev/test-cakephp/archive/refs/tags/5.0.1.1.tar.gz",
```
Over there, there's currently a 5.0.1.1 that is "latest" and another one with that name that's a draft. I assume delete the draft?

ajibarra replied:
Correct. I have already deleted the draft one

rfay reviewed (approved):
This looks great, and it can go in, thanks so much!
Two questions/issues that could be resolved if you have time/inclination:
1.  Would it be possible to add a static URL (static file) in the test setup as per comment?
2. Currently the project type is not setting upload_dirs, but for macOS users if there are in fact user-generated files in a predictable place, it would be great to set that. For example, the key items are always in sites/default/files. However, Craft CMS they're never in a predictable place really, so can't be done well there.  As it is, though, we get the warning on start on macOS and traditional Windows:
```
You have Mutagen enabled and your 'cakephp' project type doesn't have `upload_dirs` set.
For faster startup and less disk usage, set upload_dirs to where your user-generated files are stored.
If this is intended you can disable this warning with `ddev config --disable-upload-dirs-warning`.
```

tyler36 commented on docs/content/users/quickstart.md:
````diff
@@ -47,6 +47,35 @@ To get started with [Backdrop](https://backdropcms.org), clone the project repos
+
+=== "Composer"
+
+    ```bash
+    mkdir my-cakephp-app
+    cd my-cakephp-app
+    ddev config --project-type=cakephp --docroot=webroot
+    ddev composer create --prefer-dist cakephp/app:~5.0
+    ddev cake
+    ddev launch
````
When I ran through this locally, the installation stop until I answered a prompt:

```
$ddev composer create --prefer-dist cakephp/app:~5.0
...
Created `/var/www/html/tmp/cache/views` directory
Set Folder Permissions ? (Default to Y) [Y,n]? 
```

Is there a way we can avoid this?

rfay replied:
I saw the same, but this is the composer plugin requesting perms. Maybe there's an arg to `ddev composer create --prefer-dist cakephp/app:~5.0` or maybe there's a `ddev composer config` that we can do in advance?

tyler36 replied:
I tried the `-n` (--no-interaction) but it still wanted me to interact. :(
Seems like a "composer" issue and not a DDEV one so not a blocker.

rpkoller replied:
isn't that the installer.php file from cakephp that is causing the prompt?

tyler36 replied:
Digging deeper .... looks like its `App\Console\Installer.php:65` calling `static::setFolderPermissions($rootDir, $io);`

ajibarra replied:
That questions happens in postInstall so every time you execute `composer install` it asks to reset permissions. 

if you execute `composer -n install` it actually reset permissions without interaction which is expected.

However it looks like `ddev composer create` do not keep the "no interaction" when composer run-script is called (See [composer-create.go](https://github.com/ddev/ddev/blob/96fd091844b9a9cf4b09196fe0963231ed1170a4/cmd/ddev/cmd/composer-create.go#L171)). 

I was doing some tests trying to get the flag and pass it but I am probably doing something wrong because even passing `--preserve-flags`, this returns false `preserveFlags, _ := cmd.Flags().GetBool("preserve-flags")` so I was not able to deal with --no-interaction flag and pass it to `run-script`

ajibarra replied:
@rfay mentioning you here in case you want to create a ticket to investigate my findings on this.

ajibarra commented:
> This looks great, and it can go in, thanks so much! Two questions/issues that could be resolved if you have time/inclination:
> 
> 1. Would it be possible to add a static URL (static file) in the test setup as per comment?
> 2. Currently the project type is not setting upload_dirs, but for macOS users if there are in fact user-generated files in a predictable place, it would be great to set that. For example, the key items are always in sites/default/files. However, Craft CMS they're never in a predictable place really, so can't be done well there.  As it is, though, we get the warning on start on macOS and traditional Windows:
> 
> ```
> You have Mutagen enabled and your 'cakephp' project type doesn't have `upload_dirs` set.
> For faster startup and less disk usage, set upload_dirs to where your user-generated files are stored.
> If this is intended you can disable this warning with `ddev config --disable-upload-dirs-warning`.
> ```

1. I have pushed the fix in test app and test case
2. I am not sure what is the actual role of `upload_dirs`. Could you please give me an example? . I undertstood it was related with the files a user can upload (using a form or whatever) but in this case it is up to the user. If it is related to generated files, In cakephp generated files can be into config, src, templates or even webroot..or maybe I am misunderstanding what this setting does.

ajibarra commented:
> `upload_dirs` tells where user-generated files go, see https://ddev.readthedocs.io/en/stable/users/configuration/config/#upload_dirs
> 
> It allows `ddev import-files` to work (against the first one)
> 
> And for users on macOS and Windows, where mutagen is enabled by default, those directories are bind-mounted from the host instead of being synced via mutagen. This is a big deal on projects with a lot of files that aren't PHP files (which is all that mutagen really needs to sync).

I see..In that case as there is not a default I think we cannot set it..

rfay commented:
Thanks. How are user-generated files managed (like images and such) in CakePHP? 

It's possible to consider adding `disable_upload_dirs_warning: true` to the recipe, but it's best for people on macOS especially to know about this, because the overhead for mutagen on sites with a huge set of files can be enormous.

ajibarra commented:
> Thanks. How are user-generated files managed (like images and such) in CakePHP?
> 
> It's possible to consider adding `disable_upload_dirs_warning: true` to the recipe, but it's best for people on macOS especially to know about this, because the overhead for mutagen on sites with a huge set of files can be enormous.

CakePHP does not enforce any location for user uploaded files. It is up to the developer.  

Yeah I will disable upload dirs warning.

rfay commented:
Just for interest's sake, in the CakePHP community is there a "more common" local development machine/OS? Macs seem to be most common in the Drupal world, WSL2 quite common in TYPO3.

ajibarra commented:
> Just for interest's sake, in the CakePHP community is there a "more common" local development machine/OS? Macs seem to be most common in the Drupal world, WSL2 quite common in TYPO3.

Interesting question...I dont have that stats for CakePHP but at CakeDC we have around 35% Linux 35% Windows and 30% macs..

rfay commented:
YAY, congratulations and thanks! And thanks for your future maintenance of this!

ajibarra commented:
> Please let us know if CakePHP is going to make DDEV "official", always fun to know.

Since there is not an official way to install what I will do is to add a new section into the installation page in book to encourage developers to use ddev since it is clearly the easier way to do it  with the same steps we already have here.

https://book.cakephp.org/5/en/installation.html

ajibarra commented:
@rfay one last question. When is supposed to be the next release? We want to prepare a blog post as well on CakeDC blog

rfay commented:
I hope to have DDEV v1.23.0 within a very few weeks. Pre-release with cakephp in it today, it will. be v1.23.0-alpha1
